types/src/shared/cache.ts

```diff
@@ -41,26 +41,29 @@ export function cacheWithRedis<T extends (...args: any[]) => Promise<any>>(
         return JSON.parse(cacheVal) as Awaited<ReturnType<T>>;
       }
 
-      if (lockCaching) {
-        // if value not found, lock, recheck and set
-        // we avoid locking for the first read to allow parallel calls to redis if the value is set
-        await lock(key);
-        cacheVal = await redisCli.get(key);
-        if (cacheVal) {
-          return JSON.parse(cacheVal) as Awaited<ReturnType<T>>;
+      // specific try-finally to ensure unlock is called only after lock
+      try {
+        if (lockCaching) {
+          // if value not found, lock, recheck and set
+          // we avoid locking for the first read to allow parallel calls to redis if the value is set
+          await lock(key);
+          cacheVal = await redisCli.get(key);
+          if (cacheVal) {
+            return JSON.parse(cacheVal) as Awaited<ReturnType<T>>;
+          }
         }
-      }
 
-      const result = await fn(...args);
-      await redisCli.set(key, JSON.stringify(result), {
-        PX: ttlMs,
-      });
-      return result;
-    } finally {
-      if (lockCaching) {
-        unlock(key);
+        const result = await fn(...args);
+        await redisCli.set(key, JSON.stringify(result), {
+          PX: ttlMs,
+        });
+        return result;
+      } finally {
+        if (lockCaching) {
+          unlock(key);
+        }
       }
-
+    } finally {
```

Comment:
This is not catching any error, how does it solves the problem ?

Reply:
The problem was the `unlock` call inside the finally having been called without a previous `lock` call
This is because the try / catch was too wide so there could be an error in the original try/finally block before lock was called --- in which case the unlock in the finally would fail
Hope I'm clear, otherwise we can take 5' IRL :+1:

Reply:
ok all clear now ! can you move the whole block (45-65) to a dedicated method to avoid having the nested try/finally 🙏 ?

Reply:
I thought about that, but it's hard to do since part of it is specific to caching and part of it isn't and is the natural continuation of the above
I could still do it but it would be a bit artificial, not sure it would help readability